mailbox/plugin/deleted-messages-vault-postgres/src/main/java/org/apache/james/vault/metadata/PostgresDeletedMessageMetadataVault.java

```diff
@@ -0,0 +1,115 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault.metadata;
+
+import static org.apache.james.util.ReactorUtils.publishIfPresent;
+import static org.apache.james.vault.metadata.PostgresDeletedMessageMetadataModule.DeletedMessageMetadataTable.BLOB_ID;
+import static org.apache.james.vault.metadata.PostgresDeletedMessageMetadataModule.DeletedMessageMetadataTable.BUCKET_NAME;
+import static org.apache.james.vault.metadata.PostgresDeletedMessageMetadataModule.DeletedMessageMetadataTable.MESSAGE_ID;
+import static org.apache.james.vault.metadata.PostgresDeletedMessageMetadataModule.DeletedMessageMetadataTable.METADATA;
+import static org.apache.james.vault.metadata.PostgresDeletedMessageMetadataModule.DeletedMessageMetadataTable.OWNER;
+import static org.apache.james.vault.metadata.PostgresDeletedMessageMetadataModule.DeletedMessageMetadataTable.TABLE_NAME;
+import static org.jooq.JSONB.jsonb;
+
+import java.util.function.Function;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.core.Username;
+import org.apache.james.mailbox.model.MessageId;
+import org.jooq.Record;
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresDeletedMessageMetadataVault implements DeletedMessageMetadataVault {
+    private final PostgresExecutor postgresExecutor;
+    private final MetadataSerializer metadataSerializer;
+    private final BlobId.Factory blobIdFactory;
+
+    @Inject
+    public PostgresDeletedMessageMetadataVault(PostgresExecutor postgresExecutor,
+                                               MetadataSerializer metadataSerializer,
+                                               BlobId.Factory blobIdFactory) {
+        this.postgresExecutor = postgresExecutor;
+        this.metadataSerializer = metadataSerializer;
+        this.blobIdFactory = blobIdFactory;
+    }
+
+    @Override
+    public Publisher<Void> store(DeletedMessageWithStorageInformation deletedMessage) {
+        return postgresExecutor.executeVoid(context -> Mono.from(context.insertInto(TABLE_NAME)
+            .set(OWNER, deletedMessage.getDeletedMessage().getOwner().asString())
+            .set(MESSAGE_ID, deletedMessage.getDeletedMessage().getMessageId().serialize())
+            .set(BUCKET_NAME, deletedMessage.getStorageInformation().getBucketName().asString())
+            .set(BLOB_ID, deletedMessage.getStorageInformation().getBlobId().asString())
+            .set(METADATA, jsonb(metadataSerializer.serialize(deletedMessage)))));
```

Comment:
We are having a POJO that we convert into a JSON tree that we convert into a String that we parse again (jsonb) as a JSON tree...

I wonder if it would be doable to bypass the intermediary string representation if for instance using things like `jooq-jackson-extensions`...

Maybe not that significant here but it is worth a look...

Reply:
I tried the `jooq-jackson-extensions` but it seems to convert Java object to String just like what we are doing, nothing more special.

![image](https://github.com/apache/james-project/assets/55171818/86c2e304-4997-4b17-87be-e8ef9fa118f2)

> that we convert into a String that we parse again (jsonb) as a JSON tree...

BTW `JSONB` class seems to be just a string wrapper, and SQL query to insert jsonb just using a normal text value also. So likely no `parse again (jsonb) as a JSON tree` part.